### DISCUSSION TITLE
huge memory usage on scan with a lot of templates

### DISCUSSION THREAD

**Discussion by ghost** [projectdiscovery/nuclei#4626] [Q&A]
### Nuclei version:

v3.1.3 and v3.1.4-dev

### Current Behavior:

huge memory usage when nuclei scan is started with a lot of templates.
We currently have >100k templates.
The VM running nuclei freezes shortly after nuclei command is executed.

### Expected Behavior:

a plugin based scanner should not die when a lot of plugins are used.

### Steps To Reproduce:

1. Run nuclei with a lot of templates
2. see huge memory usage and VM freeze

**Comment by tarunKoyalwar** [comment]
@chrisusdag , thanks for reporting the issue but memory usage depends on lot of factors. nuclei loads all templates in memory to attempt loading with filters and other preprocessing but even loading 100k templates in-memory would only take MB's of RAM (maybe 100-500MB) . 

the core issue of VM freezing and huge memory depends on 
- config (i.e concurrency & bulk-size settings ) = more concurrency more data in memory ~ high memory usage
- template quality ( in recent test https://github.com/projectdiscovery/nuclei/issues/4552) . we ran all official nuclei-templates from projectdiscovery/nuclei-templates repo and you can see memory usage we found was `700MB - 1.2GB`.  so if template is not optimized when writing (ex: more requests in single template etc just best practices) it could lead to more memory usage and even some crash nuclei by exhausting stack( in case of intentional/ incorrect use in javascript protocol templates )
- resource intensive protocols like headless


you can try running nuclei in profiling mode (see https://github.com/projectdiscovery/nuclei/issues/4552 for ref) and share last 10-20 profiles . that way we have more info on what's causing issue and hopefully fix it soon

more docs on how to generate memory / cpu profiles https://github.com/projectdiscovery/utils/tree/main/pprof

Thank you
